[AWS Catalog] Handle the exceptions when the user does not have access to the region

Michaelvll opened the pull request:
<!-- Describe the changes in this PR -->

Describe the changes in this PR:

Fixes #1459 

Our previous `fetch_availability_zone_mappings` does not handle the case where the user does not have access to some of the regions, causing the problem in #1459, where the @iojw fails to fetch his own mapping.

**The previous problem could affect all the AWS users without the `ap-east-1` access.**

It also adds the integrity check to make sure that the auto-update of the catalog will always fetch all the regions.


<!-- Describe tests ran in a "Tested:" section -->
<!-- Unit tests (tests/test_*.py) are part of Github CI; the below are additional tests. -->

Tested (run the relevant ones):

- [x] `python -m sky.clouds.service_catalog.data_fetchers.fetch_aws --strict-all-regions`
- [x] `python -m sky.clouds.service_catalog.data_fetchers.fetch_aws` using an account without access to all regions
- [x] `sky launch -c test-mapping --cloud aws` on an account without access to all regions
- [x] @iojw, could you test with this PR to see if this will correctly generate the `az_mapping` for your account?

concretevitamin commented on sky/clouds/service_catalog/data_fetchers/fetch_aws.py:
```diff
@@ -69,10 +72,14 @@ def _get_instance_types(region: str) -> pd.DataFrame:
+        zones = []
+        response = client.describe_availability_zones()
+    except aws.client_exception():
+        # The user's AWS account may not have access to this region.
```
nit: Maybe add a comment of an example exception message?

Michaelvll replied:
Good point! Added. Thanks!

concretevitamin commented on sky/clouds/service_catalog/data_fetchers/fetch_aws.py:
```diff
@@ -69,10 +72,14 @@ def _get_instance_types(region: str) -> pd.DataFrame:
+def _get_availability_zones(region: str) -> Optional[pd.DataFrame]:
+    try:
+        client = aws.client('ec2', region_name=region)
+        zones = []
```
nit: move outside

concretevitamin commented on sky/clouds/service_catalog/data_fetchers/fetch_aws.py:
```diff
@@ -5,36 +5,39 @@
 from sky.adaptors import aws
 
-# Turn off the regions disabled for a new AWS account by default.
+# Enable all the regions. The regions enabled by the user's AWS account will be
+# filtered by joining with the availability zone mapping in `aws_catalog`.
```
```suggestion
# Enable most of the regions. Each user's account may have a subset of these enabled; this is ok because we take the intersection of the list here with the user-specific enabled regions in `aws_catalog`, via the "availability zone mapping".
```

concretevitamin commented on sky/clouds/service_catalog/data_fetchers/fetch_aws.py:
```diff
@@ -69,10 +72,14 @@ def _get_instance_types(region: str) -> pd.DataFrame:
+    try:
+        client = aws.client('ec2', region_name=region)
+        zones = []
+        response = client.describe_availability_zones()
```
Should we only try-catch this? https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/boto3.html & @iojw suggest `aws.client()` above doesn't raise.

Michaelvll replied:
Good point! Moved. Thanks!

concretevitamin commented on sky/clouds/service_catalog/data_fetchers/fetch_aws.py:
```diff
@@ -320,18 +334,35 @@ def fetch_availability_zone_mappings() -> pd.DataFrame:
+    def _check_regions_integrity(df: pd.DataFrame, name: str):
+        fetched_regions = set(df['Region'].unique())
+        requested_regions = set(region_filter)
+        if fetched_regions != requested_regions:
```
When will this happen? Worth a comment?

Michaelvll replied:
It can happen if we fail to fetch any of the regions in `region_filter`, due to network issues, AWS API glitch, etc.

concretevitamin replied:
I see. If users manually run this fetcher, is it likely such errors happen to them (I've seen some IncompleteRead)? If so, is not checking okay? Asking as I'm thinking whether this flag should be removed and always default to yes.

Michaelvll replied:
Good point! I changed logic of the fetch_aws:
1. We check the fetched regions contains all the regions enabled with the user's account by default
2. We always fetch all the regions, instead of the US regions
3. If `--strict-all-regions` is specified, we will check if the fetched regions match the `ALL_REGIONS` list despite the enabled regions of the user's account.
PTAL.

concretevitamin commented on sky/clouds/service_catalog/data_fetchers/fetch_aws.py:
```diff
@@ -320,18 +334,35 @@ def fetch_availability_zone_mappings() -> pd.DataFrame:
                         action='store_false')
+    parser.add_argument('--check-regions-integrity',
+                        action='store_true',
+                        help='Check the region integrity of the fetched data.')
```
At first glance not sure what "region integrity" means.

Michaelvll replied:
Fixed the helpstr. How does that look?

concretevitamin reviewed (commented):
Looks good! A few comments.

concretevitamin commented on sky/clouds/service_catalog/data_fetchers/fetch_aws.py:
```diff
@@ -334,9 +341,11 @@ def fetch_availability_zone_mappings() -> pd.DataFrame:
+    parser.add_argument(
+        '--check-regions-match-request',
+        action='store_true',
+        help=('Check whether the fetched data contains all regions '
+              'requested.'))
```
```suggestion
        help=('Check whether the fetched data contains all regions '
              'requested, to guard against network issues/glitches.'))
```

How about call it `--strict` or `--check-fetched-results`? IMO it seems a bit easier to parse.

concretevitamin reviewed (approved):
Thanks @Michaelvll! Some minor comments.

Michaelvll commented:
Thank you for the review @concretevitamin! It is a great point that the checking is a bit confusing and not by default. I changed the logic of the fetching, so that it will always fetch all the regions enabled by the user. I also changed the meaning the `--check` option (renamed to `--strict-all-regions`), which enforces that the regions fetched should be the same as all the regions. PTAL.

concretevitamin commented on sky/clouds/service_catalog/data_fetchers/fetch_aws.py:
```diff
@@ -60,6 +58,18 @@
+        aws_client = aws.client('ec2', region_name='us-east-1')
+        regions_enabled = aws_client.describe_regions()['Regions']
+        regions_enabled = {r['RegionName'] for r in regions_enabled}
+        regions_enabled = regions_enabled.intersection(set(ALL_REGIONS))
```
Does the github runner's credentials enable all regions in ALL_REGIONS?

Michaelvll replied:
Yes, it is an IAM role in my account that enables all the regions.

concretevitamin commented on sky/clouds/service_catalog/data_fetchers/fetch_aws.py:
```diff
@@ -342,39 +357,43 @@ def fetch_availability_zone_mappings() -> pd.DataFrame:
-        help=('Check whether the fetched data contains all regions '
-              'requested.'))
+        help=('Enforce the fetched data to contains all regions requested (all '
+              'regions or US regions).'))
```
How about `--check-all-regions-enabled-for-account` (ok to be long since it's used in github actions only)?

Help: Check that this account has enabled "all" global regions hardcoded in this script. Useful to ensure our automatic fetcher fetches the expected data.

concretevitamin reviewed (commented):
Thanks @Michaelvll. A few questions.

concretevitamin commented on mypy-cli.out:
```diff
@@ -0,0 +1,78 @@
+sky/cli.py:905: error: Incompatible return value type (got "Tuple[bool, Optional[Any]]", expected "Tuple[bool, Dict[Any, Any]]")  [return-value]
```
Nice! Remnant?

Michaelvll replied:
Oops, removed. Thanks!

concretevitamin commented on sky/clouds/service_catalog/data_fetchers/fetch_aws.py:
```diff
@@ -357,42 +359,42 @@ def fetch_availability_zone_mappings() -> pd.DataFrame:
-              'regions or US regions).'))
+        help=('Check that this account has enabled "all" global regions '
+              'hardcoded in this script. Useful to ensure our automatic '
+              'fetcher fetches the expected data.s'))
```
```suggestion
              'fetcher fetches the expected data.'))
```

concretevitamin reviewed (approved):
Thanks @Michaelvll. LGTM.
